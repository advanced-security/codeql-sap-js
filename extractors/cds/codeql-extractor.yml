name: "cds"
display_name: "SAP CAP CDS"
version: 0.0.1
column_kind: "utf16"
file_types:
  - name: cds
    display_name: SAP CAP CDS files
    extensions:
      - .cds
build_modes:
  - none