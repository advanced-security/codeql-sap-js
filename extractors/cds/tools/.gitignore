# Ignore files create just for debugging the CDS extractor.
debug/

# Override the repository-level .gitignore to explicitly include the "dist" and
#"node_modules" sub-directories, which are used to store the CDS extractor JS build
#files and their Node.js dependencies, respectively.
!dist/
!node_modules/

