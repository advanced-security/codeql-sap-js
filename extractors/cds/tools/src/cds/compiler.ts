import { execFileSync, spawnSync, SpawnSyncReturns } from 'child_process';
import { resolve } from 'path';

import { fileExists, dirExists, recursivelyRenameJsonFiles } from '../filesystem';

/**
 * Result of a CDS compilation
 */
export interface CdsCompilationResult {
  success: boolean;
  message?: string;
  outputPath?: string;
}

/**
 * Determine the `cds` command to use based on the environment.
 * @returns A string representing the CLI command to run to invoke the
 * CDS compiler.
 */
export function determineCdsCommand(): string {
  let cdsCommand = 'cds';
  // TODO : create a mapping of project sub-directories to the correct
  // cds command to use, which will also determine the version of the cds
  // compiler that will be used for compiling `.cds` files to `.cds.json`
  // files for that sub-directory / project.
  try {
    execFileSync('cds', ['--version'], { stdio: 'ignore' });
  } catch {
    // If 'cds' command is not available, use npx to run it
    cdsCommand = 'npx -y --package @sap/cds-dk cds';
  }
  return cdsCommand;
}

/**
 * Compile a CDS file to JSON
 * @param cdsFilePath Path to the CDS file
 * @param sourceRoot The source root directory
 * @param cdsCommand The CDS command to use
 * @returns Result of the compilation
 */
export function compileCdsToJson(
  cdsFilePath: string,
  sourceRoot: string,
  cdsCommand: string,
): CdsCompilationResult {
  try {
    const resolvedCdsFilePath = resolve(cdsFilePath);
    if (!fileExists(resolvedCdsFilePath)) {
      throw new Error(`Expected CDS file '${resolvedCdsFilePath}' does not exist.`);
    }

    const cdsJsonOutPath = `${resolvedCdsFilePath}.json`;
    console.log(`Processing CDS file ${resolvedCdsFilePath} to ${cdsJsonOutPath} ...`);

    const result: SpawnSyncReturns<Buffer> = spawnSync(
      cdsCommand,
      [
        'compile',
        resolvedCdsFilePath,
        '--to',
        'json',
        '--dest',
        cdsJsonOutPath,
        '--locations',
        '--log-level',
        'warn',
      ],
      { cwd: sourceRoot, shell: true, stdio: 'pipe' },
    );

    if (result.error) {
      throw new Error(`Error executing CDS compiler: ${result.error.message}`);
    }

    if (result.status !== 0) {
      throw new Error(
        `Could not compile the file ${resolvedCdsFilePath}.\nReported error(s):\n\`\`\`\n${
          result.stderr?.toString() || 'Unknown error'
        }\n\`\`\``,
      );
    }

    if (!fileExists(cdsJsonOutPath) && !dirExists(cdsJsonOutPath)) {
      throw new Error(
        `CDS source file '${resolvedCdsFilePath}' was not compiled to JSON. This is likely because the file does not exist or is not a valid CDS file.`,
      );
    }

    // Handle directory output if the CDS compiler generated a directory
    if (dirExists(cdsJsonOutPath)) {
      console.log(`CDS compiler generated JSON to output directory: ${cdsJsonOutPath}`);
      // Recursively rename all .json files to have a .cds.json extension
      recursivelyRenameJsonFiles(cdsJsonOutPath);
    } else {
      console.log(`CDS compiler generated JSON to file: ${cdsJsonOutPath}`);
    }

    return { success: true, outputPath: cdsJsonOutPath };
  } catch (error) {
    return { success: false, message: String(error) };
  }
}
