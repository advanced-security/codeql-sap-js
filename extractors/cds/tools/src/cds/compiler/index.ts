export { determineCdsCommand, getCommandAnalysisForDebug, resetCdsCommandCache } from './command';
export type { CdsCommandAnalysis } from './command';
export { compileCdsToJson } from './compile';
export {
  executeCompilationTasks,
  generateStatusReport,
  orchestrateCompilation,
  planCompilationTasks,
} from './graph';
export { findProjectForCdsFile } from './project';
export type {
  AlternativeCdsCommand,
  CdsCompilationResult,
  CompilationAttempt,
  CompilationStatus,
  CompilationTask,
  EnhancedCompilationConfig,
} from './types';
export { getCdsVersion } from './version';
