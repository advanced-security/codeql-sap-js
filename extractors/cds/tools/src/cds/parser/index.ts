export { writeParserDebugInfo } from './debug';
export { buildCdsProjectDependencyGraph, buildEnhancedCdsProjectDependencyGraph } from './graph';
export * from './functions';
export * from './types';
