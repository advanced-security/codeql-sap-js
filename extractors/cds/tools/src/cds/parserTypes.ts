/** Context for CDS access control. */
export interface CdsAccessControl {
  /** Source file where the access control is defined. */
  sourceFile: string;

  /** Service or entity being restricted. */
  target: string;

  /** Type of restriction (e.g. 'requires', 'grant'). */
  type: string;

  /** Authorization value. */
  value: unknown;
}

/** Represents a CDS annotation. */
export interface CdsAnnotation {
  /** Name of the annotation. */
  name: string;

  /** Source file where the annotation is defined. */
  sourceFile: string;

  /** Value of the annotation. */
  value: unknown;
}

/** Represents a CDS entity definition. */
export interface CdsEntity {
  /** Annotations attached to the entity. */
  annotations: CdsAnnotation[];

  /** Parent entity or aspect if extended. */
  extends?: string;

  /** Full qualified name including namespace. */
  fqn: string;

  /** Name of the entity. */
  name: string;

  /** Properties of the entity. */
  properties: CdsProperty[];

  /** Source file where the entity is defined. */
  sourceFile: string;
}

/** Represents an entity exposed in a CDS service. */
export interface CdsExposedEntity {
  /** Annotations attached to the exposed entity. */
  annotations: CdsAnnotation[];

  /** Whether this is a projection. */
  isProjection: boolean;

  /** Name of the exposed entity in the service. */
  name: string;

  /** Source entity that is being exposed. */
  sourceEntity: string;
}

/** Represents an import reference in a CDS file. */
export interface CdsImport {
  /** Whether the import is from a module (node_modules). */
  isModule: boolean;

  /** Whether the import is relative. */
  isRelative: boolean;

  /** Path to the imported resource. */
  path: string;

  /** Original import statement. */
  statement: string;
}

/** Results of parsing CDS content. */
export interface CdsParseResult {
  /** List of access controls found. */
  accessControls: CdsAccessControl[];

  /** List of context blocks. */
  contexts: {
    entities: CdsEntity[];
    name: string;
    services: CdsService[];
  }[];

  /** List of entities found. */
  entities: CdsEntity[];

  /** Errors encountered during parsing. */
  errors: string[];

  /** List of imports found. */
  imports: CdsImport[];

  /** Namespace declared in the file. */
  namespace?: string;

  /** List of services found. */
  services: CdsService[];
}

/** Represents a property of a CDS entity. */
export interface CdsProperty {
  /** Annotations attached to the property. */
  annotations: CdsAnnotation[];

  /** Cardinality if this is an association. */
  cardinality?: 'one' | 'many';

  /** Whether this property is an association. */
  isAssociation: boolean;

  /** Whether this is a composition. */
  isComposition: boolean;

  /** Whether this property is a key. */
  isKey: boolean;

  /** Name of the property. */
  name: string;

  /** Target entity if this is an association. */
  target?: string;

  /** Data type of the property. */
  type: string;
}

/** Represents a simplified package.json file structure with only the fields we need */
export interface PackageJson {
  /** The name of the package */
  name?: string;

  /** The version of the package */
  version?: string;

  /** Production dependencies */
  dependencies?: Record<string, string>;

  /** Development dependencies */
  devDependencies?: Record<string, string>;

  /** All other fields in package.json */
  [key: string]: unknown;
}

/** File cache to avoid reading the same file multiple times */
export interface FileCache {
  /** Map of file paths to their content */
  fileContents: Map<string, string>;

  /** Map of file paths to their parsed package.json content */
  packageJsonCache: Map<string, PackageJson>;

  /** Map of file paths to their parsed CDS content */
  cdsParseCache: Map<string, CdsParseResult>;
}

/** Represents a CDS project with its directory and associated files. */
export interface CdsProject {
  /** All CDS files within this project. */
  cdsFiles: string[];

  /** Dependencies on other CDS projects. */
  dependencies?: CdsProject[];

  /** The package.json content if available. */
  packageJson?: PackageJson;

  /** The directory path of the project. */
  projectDir: string;
}

/** Represents a CDS service definition. */
export interface CdsService {
  /** Annotations attached to the service. */
  annotations: CdsAnnotation[];

  /** Exposed entities in this service. */
  entities: CdsExposedEntity[];

  /** Full qualified name including namespace. */
  fqn: string;

  /** Name of the service. */
  name: string;

  /** Source file where the service is defined. */
  sourceFile: string;
}
