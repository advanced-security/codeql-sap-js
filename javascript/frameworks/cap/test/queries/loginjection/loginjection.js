import cds from '@sap/cds'
const { Books } = cds.entities('sap.capire.bookshop')

class SampleVulnService extends cds.ApplicationService {
  init() {
    // contains a sample log injection
    this.on('submitOrder', async req => {
      const { book, quantity } = req.data

      const LOG = cds.log("nodejs");
      LOG.info("CAP:" + book); // CAP log injection alert
      console.log("console:" + book); // non-CAP Log injection alert
    })

    this.on('format', (req) => {
      const cds2 = require('@sap/cds/lib')
      const LOG = cds2.log('cds.log')
      const $ = req.data; LOG.info('format:', $) // CAP log injection alert
    })

    const app = require("express")();
    app.get("search", function handler(req2, res) {
      const { book, quantity } = req2.params.category
      const LOG = cds.log("nodejs");
      LOG.info("CAP:" + book); // CAP log injection alert from non-CAP source
      console.log("console:" + book); // non-CAP Log injection alert from non-CAP source
    });

    return super.init()
  }
}
export { SampleVulnService }
