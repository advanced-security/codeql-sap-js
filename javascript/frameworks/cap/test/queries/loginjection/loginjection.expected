nodes
| loginjection.js:7:34:7:36 | req |
| loginjection.js:7:34:7:36 | req |
| loginjection.js:7:34:7:36 | req |
| loginjection.js:7:34:7:36 | req |
| loginjection.js:8:13:8:30 | { book, quantity } |
| loginjection.js:8:13:8:30 | { book, quantity } |
| loginjection.js:8:13:8:41 | book |
| loginjection.js:8:13:8:41 | book |
| loginjection.js:8:15:8:18 | book |
| loginjection.js:8:15:8:18 | book |
| loginjection.js:8:34:8:36 | req |
| loginjection.js:8:34:8:36 | req |
| loginjection.js:8:34:8:41 | req.data |
| loginjection.js:8:34:8:41 | req.data |
| loginjection.js:11:16:11:28 | "CAP:" + book |
| loginjection.js:11:16:11:28 | "CAP:" + book |
| loginjection.js:11:25:11:28 | book |
| loginjection.js:12:19:12:35 | "console:" + book |
| loginjection.js:12:19:12:35 | "console:" + book |
| loginjection.js:12:32:12:35 | book |
| loginjection.js:15:24:15:26 | req |
| loginjection.js:15:24:15:26 | req |
| loginjection.js:18:13:18:24 | $ |
| loginjection.js:18:17:18:19 | req |
| loginjection.js:18:17:18:24 | req.data |
| loginjection.js:18:47:18:47 | $ |
| loginjection.js:18:47:18:47 | $ |
| loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:23:13:23:53 | book |
| loginjection.js:23:13:23:53 | book |
| loginjection.js:23:15:23:18 | book |
| loginjection.js:23:15:23:18 | book |
| loginjection.js:23:34:23:53 | req2.params.category |
| loginjection.js:23:34:23:53 | req2.params.category |
| loginjection.js:23:34:23:53 | req2.params.category |
| loginjection.js:23:34:23:53 | req2.params.category |
| loginjection.js:25:16:25:28 | "CAP:" + book |
| loginjection.js:25:16:25:28 | "CAP:" + book |
| loginjection.js:25:25:25:28 | book |
| loginjection.js:26:19:26:35 | "console:" + book |
| loginjection.js:26:19:26:35 | "console:" + book |
| loginjection.js:26:32:26:35 | book |
edges
| loginjection.js:7:34:7:36 | req | loginjection.js:8:34:8:36 | req |
| loginjection.js:7:34:7:36 | req | loginjection.js:8:34:8:36 | req |
| loginjection.js:7:34:7:36 | req | loginjection.js:8:34:8:36 | req |
| loginjection.js:7:34:7:36 | req | loginjection.js:8:34:8:36 | req |
| loginjection.js:8:13:8:30 | { book, quantity } | loginjection.js:8:15:8:18 | book |
| loginjection.js:8:13:8:30 | { book, quantity } | loginjection.js:8:15:8:18 | book |
| loginjection.js:8:13:8:41 | book | loginjection.js:11:25:11:28 | book |
| loginjection.js:8:13:8:41 | book | loginjection.js:12:32:12:35 | book |
| loginjection.js:8:15:8:18 | book | loginjection.js:8:13:8:41 | book |
| loginjection.js:8:15:8:18 | book | loginjection.js:8:13:8:41 | book |
| loginjection.js:8:34:8:36 | req | loginjection.js:8:34:8:41 | req.data |
| loginjection.js:8:34:8:36 | req | loginjection.js:8:34:8:41 | req.data |
| loginjection.js:8:34:8:41 | req.data | loginjection.js:8:13:8:30 | { book, quantity } |
| loginjection.js:8:34:8:41 | req.data | loginjection.js:8:13:8:30 | { book, quantity } |
| loginjection.js:11:25:11:28 | book | loginjection.js:11:16:11:28 | "CAP:" + book |
| loginjection.js:11:25:11:28 | book | loginjection.js:11:16:11:28 | "CAP:" + book |
| loginjection.js:12:32:12:35 | book | loginjection.js:12:19:12:35 | "console:" + book |
| loginjection.js:12:32:12:35 | book | loginjection.js:12:19:12:35 | "console:" + book |
| loginjection.js:15:24:15:26 | req | loginjection.js:18:17:18:19 | req |
| loginjection.js:15:24:15:26 | req | loginjection.js:18:17:18:19 | req |
| loginjection.js:18:13:18:24 | $ | loginjection.js:18:47:18:47 | $ |
| loginjection.js:18:13:18:24 | $ | loginjection.js:18:47:18:47 | $ |
| loginjection.js:18:17:18:19 | req | loginjection.js:18:17:18:24 | req.data |
| loginjection.js:18:17:18:24 | req.data | loginjection.js:18:13:18:24 | $ |
| loginjection.js:23:13:23:30 | { book, quantity } | loginjection.js:23:15:23:18 | book |
| loginjection.js:23:13:23:30 | { book, quantity } | loginjection.js:23:15:23:18 | book |
| loginjection.js:23:13:23:53 | book | loginjection.js:25:25:25:28 | book |
| loginjection.js:23:13:23:53 | book | loginjection.js:26:32:26:35 | book |
| loginjection.js:23:15:23:18 | book | loginjection.js:23:13:23:53 | book |
| loginjection.js:23:15:23:18 | book | loginjection.js:23:13:23:53 | book |
| loginjection.js:23:34:23:53 | req2.params.category | loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:23:34:23:53 | req2.params.category | loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:23:34:23:53 | req2.params.category | loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:23:34:23:53 | req2.params.category | loginjection.js:23:13:23:30 | { book, quantity } |
| loginjection.js:25:25:25:28 | book | loginjection.js:25:16:25:28 | "CAP:" + book |
| loginjection.js:25:25:25:28 | book | loginjection.js:25:16:25:28 | "CAP:" + book |
| loginjection.js:26:32:26:35 | book | loginjection.js:26:19:26:35 | "console:" + book |
| loginjection.js:26:32:26:35 | book | loginjection.js:26:19:26:35 | "console:" + book |
#select
| loginjection.js:11:16:11:28 | "CAP:" + book | loginjection.js:7:34:7:36 | req | loginjection.js:11:16:11:28 | "CAP:" + book | Log entry depends on a $@. | loginjection.js:7:34:7:36 | req | user-provided value |
| loginjection.js:18:47:18:47 | $ | loginjection.js:15:24:15:26 | req | loginjection.js:18:47:18:47 | $ | Log entry depends on a $@. | loginjection.js:15:24:15:26 | req | user-provided value |
| loginjection.js:25:16:25:28 | "CAP:" + book | loginjection.js:23:34:23:53 | req2.params.category | loginjection.js:25:16:25:28 | "CAP:" + book | Log entry depends on a $@. | loginjection.js:23:34:23:53 | req2.params.category | user-provided value |
